libraries/AP_RCProtocol/AP_RCProtocol.h

```diff
@@ -197,28 +186,33 @@ class AP_RCProtocol {
     // return true if a specific protocol is enabled
     bool protocol_enabled(enum rcprotocol_t protocol) const;
 
+    void add_input(enum rcprotocol_t protocol, uint8_t num_channels, uint16_t *values, int16_t rssi, int16_t rx_link_quality);
+
     enum rcprotocol_t _detected_protocol = NONE;
     uint16_t _disabled_for_pulses;
     bool _detected_with_bytes;
     AP_RCProtocol_Backend *backend[NONE];
     bool _new_input;
-    uint32_t _last_input_ms;
+    uint32_t _last_input_us;
     bool _failsafe_active;
     bool _valid_serial_prot;
 
     // optional additional uart
     struct {
+        SoftSerial ss;
```

Comment:
This isn't a blocker, just unfortunate.

I was looking forward to allowing a user to configure multiple backends at the same time, with different input sources, as part of my RC-over-DroneCAN work.  So allowing a user to have an RC receiver on each wing, for example.  Having soft-serial out the front here moves us away from that, as multiple SoftSerial instances would be required to allow for more (e.g.) RCINT-defined softserial instances.

Ideally, I think, the HAL should be bearing a UARTDriver driven by SoftSerial, defaulting to 8N1 but allowing for different configurations to be specified.  That would mean your `apply_to_softserial` method (called directly after `apply_to_uart`) would disappear - the `apply_to_uart` woulld affect the softserial HAL UARTDriver.  I don't know how that works in with *also* trying o determine if the pulse train is PPMSum :-)

Reply:
Yes, agreed. Note that the original PR made SoftSerial static so I think moving a very long way from what you want. Putting it as an instance in AP_RCProtocol and fixing the fallout does at least make what you want foreseeable. I also agree that SoftSerial shoud be buried in the protocol layer so that this module doesn't have to worry about it - I think my change takes us closer towards that - backends now don't really have to worry about SoftSerial (unless they are PPMSum) which is a Good Thing IMO.